reps/2023-06-30-serve-cpp-deployment.md

```diff
@@ -0,0 +1,413 @@
+Ray Serve maintains Controller Actor and Ingress Actor. So these two roles are not related to the language of the user's choice. And they have the ability to manage cross-language deployments and route requests.
+### C++ Case Deduction
+The businesses can send control commands to the Controller Actor of Ray Serve, which includes creating ingress, creating deployment, etc. When publishing C++ online services, the DeploymentState component needs to create C++ Deployment  Actors. Users can call their business logic in Python/Java Driver, Ray task, or Ray Actor, and the requests will be dispatched to C++ Deployment Actors.
+![image](https://user-images.githubusercontent.com/40862361/250100685-96466519-1f8e-414e-8d8a-1bcb63952a41.png)
+### Package
+C++ programs are typically compiled and packaged into three types of results: binary, static library, shared library.  
+- binary: C++ does not have a standard load binary API, and binaries compiled by different compilers may not be the same. Directly loading the binary can result in many uncontrollable factors.   
+- static library: Static library bundle all their dependencies together, whereas common utility libraries such as glog and boost may be loaded via dynamic dependencies in the Ray Serve Deployment. This can lead to a higher probability of conflicts when the same library exists as both a dynamic and static dependency.  
+- shared library: C++ provides a standard API for loading shared library, and it can reduce memory usage and prevent conflicts.  
+
+In conclusion, the business needs to package the system as a shared library to run it on Ray Serve.
+### Register function
+Ray Serve will add SERVE_FUNC and SERVE_DEPLOYMENT macros to publish user Service as Deployment.  
+SERVE_FUNC: Resolving overloaded function registration;
+SERVE_DEPLOYMENT: Publishing user Service as Deployment. 
+Example:
+```cpp
+static RecommendService *CreateRecommendService(std::string request) {
+  return new RecommendService(request);
+}
+
+static RecommendService *CreateRecommendService(std::string request, int num) {
+  return new RecommendService(request, num);
+}
+
+static FeatureService *CreateFeatureService() {
+  return new FeatureService();
+}
+
+// Register function
+SERVE_DEPLOYMENT(SERVE_FUNC(CreateRecommendService, std::string, int),
+			    CreateFeatureService);
+```
+## Compatibility, Deprecation, and Migration Plan
+The new feature is to add C++ deployment for Ray Serve, without modifying or deprecating existing functionalities. The changes to the Ray Serve API are also adding new capabilities.
+
+## Test Plan and Acceptance Criteria
+- Unit and integration test for core components
+- Benchmarks on C++ Deployment
+
+## (Optional) Follow-on Work
+- Init Ray Serve C++ project structure
+- Create C++ Deployment through Python/Java API
+- ServeHandler for C++
+- Accessing C++ Deployment using Python/Java/C++ ServeHandle or HTTP
```

Comment:
From my perspective I think supporting HTTP is a P0 requirement in order to really say that we support C++ as a language in Ray Serve.

Could you at least sketch a possible API that we could support here? It does not have to be part of the initial scope of implementation. But without having this I think the proposal is incomplete.

Reply:
Yes. 
Expose two primary APIs to users: SERVE_FUNC and SERVE_DEPLOYMENT, with a detailed explanation provided in the section titled "Register Function". 
Furthermore, users can deploy services through these APIs, as exemplified in the "Converting to a Ray Serve Deployment" section.

Reply:
I am a bit confused -- how would a C++ deployment be directly exposed via HTTP? Or are you proposing that in order to expose it via HTTP it must be placed behind a Python deployment that calls into it via `DeploymentHandle`?

Reply:
In the example above, I see a `RecommendService` but this takes as arguments a `std::string` and an `int`. How are these parsed from an HTTP request? And how can an HTTP response be output?

Reply:
The way of accessing via HTTP needs to be adjusted for this demo.

```
curl http://127.0.0.1:8000?request=computer&num=1
```
This is not sufficient because the type information for `num=1` is lost in the http query. For cases with multiple input parameters, it is possible to transmit them as a JSON string in the HTTP body, for example:

```
curl -d '{"request": "computer", "num": 1}' http://127.0.0.1:8000
```
This way, the request flow is: -> `HTTPProxy (access with DeploymentHandle)` -> `Cpp Deployment`.
And currently, the `HTTPProxy` should be able to meet the requirements without any additional modifications.

@MingYueQ can adjust according to this approach.